vendor/zkllvm-metacraft-circuits/src/circuit_utils/circuit_byte_utils.h

```diff
@@ -0,0 +1,187 @@
+#pragma once
+
+#include <nil/crypto3/hash/algorithm/hash.hpp>
+#include <nil/crypto3/hash/sha2.hpp>
+#include "base_types.h"
+#include "../utils/picosha2.h"
+
+using namespace nil::crypto3;
+
+namespace circuit_byte_utils {
+
+#define countof(array) (sizeof(array) / sizeof(array[0]))
+
+    using sha256_t = typename hashes::sha2<256>::block_type;
+
+    bool sha256_equals(sha256_t hash1, sha256_t hash2) {
+
+        bool result = true;
+        for (auto i = 0; i < countof(hash1); ++i) {
+            result = result && (hash1[i] == hash2[i]);
+        }
+
+        return result;
+    }
+
+    template<typename T>
+    Byte get_nth_byte(const T val, unsigned int n) {
+        static_assert(std::is_integral<typename std::remove_reference<T>::type>::value, "T must be integral");
+        assert_true(n < sizeof(T));
+
+        return T(val >> (n * 8));
+    }
+
+    bool get_nth_bit(uint64_t value, short i) {
+        return bool(1 & (value >> i));
+    }
+
+    Byte set_nth_bit(Byte& value, short i) {
+        assert_true(i < 8);
+        value = value | (Byte(1) << i);
+        return value;
+    }
+
+    template<typename T>
+    void sha256_to_bytes_array(sha256_t sha, T& out) {
+        assert_true(out.size() >= sizeof(sha));
+        for (int int_count = 0; int_count < sizeof(sha) / sizeof(sha[0]); int_count++) {
+
+            for (int byte_count = 0; byte_count < sizeof(sha[0]); byte_count++) {
+                out[int_count * sizeof(sha[0]) + byte_count] = get_nth_byte(sha[int_count], byte_count);
+            }
+        }
+    }
+
+    template<std::size_t N, typename T, std::size_t InputSize>
+    std::array<T, N> take_n_elements(const std::array<T, InputSize>& val) {
+        static_assert(N <= InputSize);
+        std::array<T, N> ret {};
+        std::copy(val.begin(), val.begin() + N, ret.begin());
+
+        return ret;
+    }
+
+    template<typename T, bool little_endian = true>
+    std::array<Byte, sizeof(T)> int_to_bytes(const T& paramInt) {
+        static_assert(std::is_integral_v<typename std::remove_reference_t<T>>, "T must be integral");
+        std::array<Byte, sizeof(T)> bytes {};
+        if constexpr (little_endian) {
+            for (int i = 0; i < sizeof(T); ++i) {
+                bytes[i] = (paramInt >> (i * 8));
+            }
+        } else {
+            for (int i = sizeof(T) - 1; i >= 0; i--) {
+                bytes[i] = (paramInt >> (i * 8));
+            }
+        }
+        return bytes;
+    }
+
+    template<typename T>
+    T bytes_to_int(const std::array<Byte, sizeof(T)>& paramVec, bool little_endian = true) {
```

Comment:
Change the little_endian function parameter to be a template parameter instead. You've fixed this on another branch but you could also change it here for completeness.